### PR TITLE
Add new benchmarks

### PR DESCRIPTION
**Pull request by lplewa**
- L0 benchmark will be added in separate PR (unless merge of this one will take too long ;))
- Old benchmark will be removed in separate PRs too
- Please focus more on bechmark framework - particular benchmark will be also changed in separate PRs - current set is more a demo of framework - to select good benchmark scenarios there is litte bit more research to be done.

### CODE REVIEW

File: benchmark/CMakeLists.txt [hunk 1]
```diff
@@ -55,13 +67,7 @@ function(add_umf_benchmark)
         COMMAND ${BENCH_NAME}
         WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
 
-    # Benchmark passes if it prints "PASSED" in the output, because ubench of
-    # scalable pool fails if the confidence interval exceeds maximum permitted
-    # 2.5%.
-    set_tests_properties(
```
[the remark this answers is not in the document]
Comment (lplewa) [reply benchmark/CMakeLists.txt:61]: Well - i whold have to do extra work to enable it only for ubench, when we will remove it few days later

File: benchmark/benchmark.cpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    virtual umf_memory_pool_ops_t *
+    getOps([[maybe_unused]] ::benchmark::State &state) {
+        return umfScalablePoolOps();
+    }
+    virtual void *getParams([[maybe_unused]] ::benchmark::State &state) {
+        return NULL;
+    }
+};
+
+ALLOC_BENCHMARK_TEMPLATE_DEFINE(alloc_benchmark, stdmalloc, fix_alloc_size);
```
[the remark this answers is not in the document]
Comment (lplewa) [reply benchmark/benchmark.cpp:98]: Yep - i added comment to indicate it. 

I have plan to make it little bit better - those scenarios are more a demo then actual final benchmarks.

File: benchmark/CMakeLists.txt [hunk 3]
```diff
@@ -117,7 +123,12 @@ endif()
 add_umf_benchmark(
     NAME ubench
     SRCS ubench.c
-    LIBS ${LIBS_OPTIONAL}
+    LIBDIRS ${LIB_DIRS})
```
[the remark this answers is not in the document]
Comment (lplewa) [reply benchmark/CMakeLists.txt:127]: done

File: benchmark/CMakeLists.txt [hunk 4]
```diff
@@ -1,10 +1,22 @@
-# Copyright (C) 2023 Intel Corporation
+# Copyright (C) 2023-2024 Intel Corporation
 # Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
 # In MSVC builds, there is no way to determine the actual build type during the
 # CMake configuration step. Therefore, this message is printed in all MSVC
 # builds.
```
[the remark this answers is not in the document]
Comment (lplewa) [reply benchmark/CMakeLists.txt]: done

File: benchmark/CMakeLists.txt [hunk 5]
```diff
@@ -1,10 +1,22 @@
...
+
+set(CMAKE_CXX_STANDARD 17)
```
[the remark this answers is not in the document]
Comment (lplewa) [reply benchmark/CMakeLists.txt]: done

File: benchmark/benchmark.hpp [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+#include <random>
+#include <umf/memory_pool.h>
+#include <umf/memory_provider.h>
+
+struct alloc_data {
+    void *ptr;
+    size_t size;
+};
+
+#define ALLOC_BENCHMARK_TEMPLATE_DEFINE(A, B, ...)                             \
```
[the remark this answers is not in the document]
Comment (lplewa) [reply benchmark/benchmark.hpp]: done

File: benchmark/benchmark.cpp [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        return umfProxyPoolOps();
+    }
+    virtual void *
+    getParams([[maybe_unused]] ::benchmark::State &state) override {
+        return nullptr;
+    }
+};
+
+#if (defined UMF_BUILD_LIBUMF_POOL_DISJOINT)
+
```
[the remark this answers is not in the document]
Comment (lplewa) [reply benchmark/benchmark.cpp:46]: done

File: benchmark/benchmark.cpp [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        }
+        disjoint_memory_pool_params.SlabMinSize = page_size;
+        disjoint_memory_pool_params.MaxPoolableSize = page_size * 2;
+        disjoint_memory_pool_params.Capacity = state.range(0);
+
+        disjoint_memory_pool_params.MinBucketSize = page_size;
+        return &disjoint_memory_pool_params;
+    }
+};
+
```
[the remark this answers is not in the document]
Comment (lplewa) [reply benchmark/benchmark.cpp:69]: done

File: benchmark/benchmark.hpp [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+        }
+    }
+
+    virtual umf_memory_provider_ops_t *getOps() { return nullptr; }
+    virtual void *getParams() { return nullptr; }
+};
+
+template <typename T,
+          typename =
+              std::enable_if_t<std::is_base_of<provider_interface, T>::value>>
```
Comment (bratpiorka) [inline benchmark/benchmark.hpp:93]: is this needed?

**Review by lukaszstolarczuk — APPROVED** [review]

**Review by bratpiorka — APPROVED** [review]

**Review by ldorau — CHANGES_REQUESTED** [review]
> please merge after #901 (will require rebase and update Disjoint Pool parameters)

@lplewa Please rebase and update Disjoint Pool parameters

File: benchmark/benchmark.hpp [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+
+class alloc_size_interface {
+  public:
+    virtual void SetUp([[maybe_unused]] ::benchmark::State &state,
+                       [[maybe_unused]] unsigned r) {}
+    virtual void TearDown([[maybe_unused]] ::benchmark::State &state) {}
+    virtual size_t nextSize() = 0;
+};
+
+class fix_alloc_size : public alloc_size_interface {
```
[the remark this answers is not in the document]
Comment (lplewa) [reply benchmark/benchmark.hpp:36]: Done

**Comment by lukaszstolarczuk** [comment]
@lplewa, please let us know when this PR is ready for re-review :)

File: benchmark/benchmark.hpp [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+    size_t size;
+};
+
+class uniform_alloc_size : public alloc_size_interface {
+    using distribution = std::uniform_int_distribution<int64_t>;
+
+  public:
+    unsigned SetUp(::benchmark::State &state, unsigned r) override {
+        auto min = state.range(r);
+        auto max = state.range(r + 1);
```
Comment (bratpiorka) [inline benchmark/benchmark.hpp:56]: r + 1/2? comment what this function does

File: benchmark/benchmark.hpp [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+    size_t max_allocs = 1000;
+    size_t pre_allocs = 0;
+    void SetUp(::benchmark::State &state) override {
+        if (state.thread_index() != 0) {
+            return;
+        }
+        max_allocs = state.range(0);
+        pre_allocs = state.range(1);
+
+        int r = base::alloc_size.SetUp(state, 2);
```
Comment (bratpiorka) [inline benchmark/benchmark.hpp:96]: 2? r + 1? this is magic :)

File: benchmark/benchmark.hpp [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+        base::allocator.SetUp(state, r + 1);
+        allocations.resize(state.threads());
+        iters.resize(state.threads());
+        for (auto &i : iters) {
```
Comment (bratpiorka) [inline benchmark/benchmark.hpp:100]: add empty line before "for" and comment what you are doing here

File: benchmark/benchmark.hpp [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+        auto &i = iters[tid];
+        allocations[tid][i].ptr = base::allocator.benchAlloc(s);
+        if (allocations[tid][i].ptr == NULL) {
+            state.SkipWithError("allocation failed");
+            return;
+        }
+        allocations[tid][i].size = s;
+        i++;
+        if (i >= max_allocs + pre_allocs) {
+            state.PauseTiming();
```
Comment (bratpiorka) [inline benchmark/benchmark.hpp:146]: please comment why

File: benchmark/benchmark.hpp (hunk 12 above)
Comment (lplewa) [reply benchmark/benchmark.hpp:96]: dito

File: benchmark/benchmark.hpp [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+
+        int r = base::alloc_size.SetUp(state, 2);
+        base::allocator.SetUp(state, r + 1);
+        allocations.resize(state.threads());
+        iters.resize(state.threads());
+        for (auto &i : iters) {
+            i = pre_allocs;
+        }
+        for (auto &i : allocations) {
+            i.resize(max_allocs + pre_allocs);
```
[the remark this answers is not in the document]
Comment (lplewa) [reply benchmark/benchmark.hpp:104]: ```std::vector<std::vector<alloc_data>> allocations; ```
98 initalise first vector and then we init inner vectors

File: benchmark/benchmark.cpp [hunk 16]
```diff
@@ -0,0 +1,354 @@
...
+        return "scalable_pool<" + Provider::name() + ">";
+    }
+};
+
+// Benchmarks scenarios:
+
+UMF_BENCHMARK_TEMPLATE_DEFINE(alloc_benchmark, glibc_fix, fixed_alloc_size,
+                              glibc_malloc);
+
+UMF_BENCHMARK_REGISTER_F(alloc_benchmark, glibc_fix)
```
Comment (bratpiorka) [inline benchmark/benchmark.cpp:173]: Add comment on what each parameter does. Also, what does it mean to pass Args twice? and Threads?

Comment (lplewa) [reply benchmark/benchmark.cpp:173]: Add comment on what each parameter does. - Done

 Also, what does it mean to pass Args twice? and Threads - this is documented in google benchmark docs (or header file). It does run benchmark in all possible configurations. In this case we have 4 configs created by caesarean product between arguments and threads set.

File: benchmark/benchmark.hpp (hunk 12 above)
Comment (lplewa) [reply benchmark/benchmark.hpp]: done

File: benchmark/benchmark.hpp (hunk 11 above)
Comment (lplewa) [reply benchmark/benchmark.hpp]: done

**Review by bratpiorka — APPROVED** [review]

File: benchmark/benchmark.cpp [hunk 17]
```diff
@@ -0,0 +1,358 @@
...
+#include <umf/pools/pool_proxy.h>
+#include <umf/pools/pool_scalable.h>
+#include <umf/providers/provider_level_zero.h>
+#include <umf/providers/provider_os_memory.h>
+
+#ifdef UMF_BUILD_LIBUMF_POOL_DISJOINT
+#include <umf/pools/pool_disjoint.h>
+#endif
+
+#ifdef UMF_BUILD_LIBUMF_POOL_JEMALLOC
```
Comment (ldorau) [inline benchmark/benchmark.cpp:19]: .

File: benchmark/benchmark.cpp [hunk 18]
```diff
@@ -0,0 +1,358 @@
...
+                              uniform_alloc_size,
+                              pool_allocator<disjoint_pool<os_provider>>);
+UMF_BENCHMARK_REGISTER_F(alloc_benchmark, disjoint_pool_uniform)
+    ->Args({10000, 0, 8, 64 * 1024, 8})
+    ->Threads(4)
+    ->Threads(1);*/
+
+#endif
+
+#ifdef UMF_BUILD_LIBUMF_POOL_JEMALLOC
```
Comment (ldorau) [inline benchmark/benchmark.cpp:228]: .

File: benchmark/benchmark.cpp [hunk 19]
```diff
@@ -0,0 +1,358 @@
...
+                              disjoint_pool_uniform, uniform_alloc_size,
+                              pool_allocator<disjoint_pool<os_provider>>);
+UMF_BENCHMARK_REGISTER_F(multiple_malloc_free_benchmark, disjoint_pool_uniform)
+    ->Args({10000, 0, 8, 64 * 1024, 8})
+    ->Threads(4)
+    ->Threads(1);
+*/
+#endif
+
+#ifdef UMF_BUILD_LIBUMF_POOL_JEMALLOC
```
Comment (ldorau) [inline benchmark/benchmark.cpp:321]: .

**Review by ldorau — CHANGES_REQUESTED** [review]

File: benchmark/benchmark.cpp [hunk 20]
```diff
@@ -0,0 +1,358 @@
...
+                              uniform_alloc_size,
+                              pool_allocator<jemalloc_pool<os_provider>>);
+UMF_BENCHMARK_REGISTER_F(alloc_benchmark, jemalloc_pool_uniform)
+    ->Args({10000, 0, 8, 64 * 1024, 8})
+    ->Threads(4)
+    ->Threads(1);
+
+#endif
+
+UMF_BENCHMARK_TEMPLATE_DEFINE(alloc_benchmark, scalable_pool_fix,
```
Comment (lukaszstolarczuk) [inline benchmark/benchmark.cpp:248]: shouldn't we also have `ifdef` for scalable pool enabled?

File: benchmark/benchmark.cpp [hunk 21]
```diff
@@ -0,0 +1,358 @@
...
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ */
+
+#include <benchmark/benchmark.h>
+#include <umf/pools/pool_proxy.h>
+#include <umf/pools/pool_scalable.h>
+#include <umf/providers/provider_level_zero.h>
```
Comment (lukaszstolarczuk) [inline benchmark/benchmark.cpp:12]: are we using L0?

File: benchmark/benchmark.cpp [hunk 22]
```diff
@@ -0,0 +1,358 @@
...
+
+UMF_BENCHMARK_TEMPLATE_DEFINE(alloc_benchmark, os_provider, fixed_alloc_size,
+                              provider_allocator<os_provider>);
+UMF_BENCHMARK_REGISTER_F(alloc_benchmark, os_provider)
+    ->Args({10000, 0, 4096})
+    ->Args({10000, 100000, 4096})
+    ->Threads(4)
+    ->Threads(1);
+
+UMF_BENCHMARK_TEMPLATE_DEFINE(alloc_benchmark, proxy_pool, fixed_alloc_size,
```
Comment (lukaszstolarczuk) [inline benchmark/benchmark.cpp:198]: and also proxy pool under `ifdef UMF_PROXY_LIB_ENABLED`

File: benchmark/benchmark.hpp [hunk 23]
```diff
@@ -0,0 +1,382 @@
+/*
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ */
+
+/*
+ * This file defines a benchmarking framework for evaluating memory allocation
```
Comment (lukaszstolarczuk) [inline benchmark/benchmark.hpp:10]: Kinda looks like wall of text... perhaps, please make it more concise.

I really like the flow description and "extension" section, though

File: benchmark/benchmark.hpp [hunk 24]
```diff
@@ -0,0 +1,382 @@
...
+        return {"min size", "max size", "granularity"};
+    }
+
+  private:
+    std::default_random_engine generator;
+    distribution dist;
+    size_t multiplier;
+};
+
+// This class benchmark speed of alloc() operations.
```
Comment (lukaszstolarczuk) [inline benchmark/benchmark.hpp:143]: did you mean `benchmark` -> `benchmarks` ?

File: benchmark/benchmark.hpp [hunk 25]
```diff
@@ -0,0 +1,382 @@
...
+        // unpack arguments
+        int argPos = 0;
+        max_allocs = state.range(argPos++);
+        pre_allocs = state.range(argPos++);
+        // pass rest of the arguments to "alloc_size" and "allocator"
+        argPos = base::alloc_size.SetUp(state, argPos);
+        base::allocator.SetUp(state, argPos);
+
+        // initialize allocations tracking vectors (one per thread)
+        // and iterators for this vectors.
```
Comment (lukaszstolarczuk) [inline benchmark/benchmark.hpp:168]: these vectors

File: benchmark/benchmark.cpp [hunk 26]
```diff
@@ -0,0 +1,358 @@
...
+
+// Benchmarks scenarios:
+
+UMF_BENCHMARK_TEMPLATE_DEFINE(alloc_benchmark, glibc_fix, fixed_alloc_size,
+                              glibc_malloc);
+
+// The benchmark arguments specified in Args() are, in order:
+// benchmark arguments, allocator arguments, size generator arguments.
+// The exact meaning of each argument depends on the benchmark, allocator, and size components used.
+// Refer to the 'name()' function in each component to find detailed descriptions of these arguments.
```
Comment (lukaszstolarczuk) [inline benchmark/benchmark.cpp:176]: did you mean `'argsName()' function`?

File: benchmark/benchmark.cpp (hunk 17 above)
Comment (lplewa) [reply benchmark/benchmark.cpp:19]: done

File: benchmark/benchmark.cpp (hunk 19 above)
Comment (lplewa) [reply benchmark/benchmark.cpp:321]: done

File: benchmark/benchmark.cpp (hunk 20 above)
Comment (lplewa) [reply benchmark/benchmark.cpp:248]: done

File: benchmark/benchmark.hpp (hunk 24 above)
Comment (lplewa) [reply benchmark/benchmark.hpp:143]: done

File: benchmark/benchmark.hpp (hunk 25 above)
Comment (lplewa) [reply benchmark/benchmark.hpp:168]: done

File: benchmark/benchmark.cpp (hunk 26 above)
Comment (lplewa) [reply benchmark/benchmark.cpp:176]: done

**Review by ldorau — APPROVED** [review]

**Pull request merged**